File: skills/duckdb-c-extension-function-catalog/references/generate_function_catalog.py
#!/usr/bin/env python3
"""Render generated function catalog artifacts from functions.yaml.

The manifest is stored in JSON-formatted YAML so this script can use only the
Python standard library, following the DuckHTS pattern.
"""

from __future__ import annotations

import csv
import json
import sys
from pathlib import Path


def die(message: str) -> None:
    print(message, file=sys.stderr)
    raise SystemExit(1)


def escape_md(text: str) -> str:
    return text.replace("|", "\\|").replace("\n", " ")


def load_manifest(path: Path) -> list[dict[str, object]]:
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        die(f"Failed to parse {path}: {exc}")

    functions = payload.get("functions")
    if not isinstance(functions, list):
        die(f"{path} is missing a top-level 'functions' array")
    return functions


def render_markdown(functions: list[dict[str, object]]) -> str:
    lines = [
        "# Function Catalog",
        "",
        "This file is generated from `functions.yaml`.",
        "",
        "| name | kind | returns | since | deprecated | description |",
        "|---|---|---|---|---|---|",
    ]
    for entry in functions:
        deprecated = entry.get("deprecated") or {}
        deprecated_since = deprecated.get("since", "") if isinstance(deprecated, dict) else ""
        lines.append(
            "| `{name}` | {kind} | `{returns}` | {since} | {deprecated} | {description} |".format(
                name=escape_md(str(entry.get("name", ""))),
                kind=escape_md(str(entry.get("kind", ""))),
                returns=escape_md(str(entry.get("returns", ""))),
                since=escape_md(str(entry.get("since", ""))),
                deprecated=escape_md(str(deprecated_since)),
                description=escape_md(str(entry.get("description", ""))),
            )
        )
    lines.append("")
    return "\n".join(lines)


def write_tsv(path: Path, functions: list[dict[str, object]]) -> None:
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.writer(handle, delimiter="\t", lineterminator="\n")
        writer.writerow(["name", "kind", "returns", "since", "deprecated", "description"])
        for entry in functions:
            deprecated = entry.get("deprecated") or {}
            deprecated_since = deprecated.get("since", "") if isinstance(deprecated, dict) else ""
            writer.writerow(
                [
                    entry.get("name", ""),
                    entry.get("kind", ""),
                    entry.get("returns", ""),
                    entry.get("since", ""),
                    deprecated_since,
                    str(entry.get("description", "")).replace("\n", " "),
                ]
            )


def main() -> None:
    if len(sys.argv) != 3:
        die(f"usage: {Path(sys.argv[0]).name} <functions.yaml> <outdir>")

    manifest_path = Path(sys.argv[1]).resolve()
    outdir = Path(sys.argv[2]).resolve()
    outdir.mkdir(parents=True, exist_ok=True)

    functions = load_manifest(manifest_path)
    (outdir / "functions.md").write_text(render_markdown(functions), encoding="utf-8")
    write_tsv(outdir / "functions.tsv", functions)
    print(f"wrote {outdir / 'functions.md'}")
    print(f"wrote {outdir / 'functions.tsv'}")


if __name__ == "__main__":
    main()

File: skills/duckdb-c-extension/references/generate_function_catalog.R
#!/usr/bin/env Rscript

# Render generated function catalog artifacts from functions.yaml.
#
# The manifest is expected to be JSON-formatted YAML/JSON so R can parse it
# deterministically with jsonlite. Keep this script small and project-local.

die <- function(...) {
  message(...)
  quit(status = 1, save = "no")
}

escape_md <- function(text) {
  text <- as.character(if (is.null(text)) "" else text)
  text <- gsub("\\|", "\\\\|", text)
  gsub("[\r\n]+", " ", text)
}

clean_tsv <- function(text) {
  text <- as.character(if (is.null(text)) "" else text)
  gsub("[\r\n\t]+", " ", text)
}

field <- function(entry, name, default = "") {
  value <- entry[[name]]
  if (is.null(value)) return(default)
  as.character(value)
}

deprecated_since <- function(entry) {
  deprecated <- entry[["deprecated"]]
  if (is.list(deprecated)) field(deprecated, "since") else ""
}

load_manifest <- function(path) {
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    die("Install the R package 'jsonlite' to parse ", path)
  }

  payload <- tryCatch(
    jsonlite::fromJSON(path, simplifyVector = FALSE),
    error = function(error) die("Failed to parse ", path, ": ", conditionMessage(error))
  )

  functions <- payload[["functions"]]
  if (!is.list(functions)) {
    die(path, " is missing a top-level 'functions' array")
  }
  functions
}

render_markdown <- function(functions) {
  lines <- c(
    "# Function Catalog",
    "",
    "This file is generated from `functions.yaml`.",
    "",
    "| name | kind | returns | since | deprecated | description |",
    "|---|---|---|---|---|---|"
  )

  for (entry in functions) {
    lines <- c(
      lines,
      sprintf(
        "| `%s` | %s | `%s` | %s | %s | %s |",
        escape_md(field(entry, "name")),
        escape_md(field(entry, "kind")),
        escape_md(field(entry, "returns")),
        escape_md(field(entry, "since")),
        escape_md(deprecated_since(entry)),
        escape_md(field(entry, "description"))
      )
    )
  }

  paste0(paste(lines, collapse = "\n"), "\n")
}

write_tsv <- function(path, functions) {
  header <- c("name", "kind", "returns", "since", "deprecated", "description")
  rows <- lapply(functions, function(entry) {
    c(
      field(entry, "name"),
      field(entry, "kind"),
      field(entry, "returns"),
      field(entry, "since"),
      deprecated_since(entry),
      field(entry, "description")
    )
  })

  lines <- c(
    paste(header, collapse = "\t"),
    vapply(rows, function(row) paste(clean_tsv(row), collapse = "\t"), character(1))
  )
  writeLines(lines, path, useBytes = TRUE)
}

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2) {
  die("usage: generate_function_catalog.R <functions.yaml> <outdir>")
}

manifest_path <- normalizePath(args[[1]], winslash = "/", mustWork = TRUE)
outdir <- normalizePath(args[[2]], winslash = "/", mustWork = FALSE)
dir.create(outdir, recursive = TRUE, showWarnings = FALSE)

functions <- load_manifest(manifest_path)
md_path <- file.path(outdir, "functions.md")
tsv_path <- file.path(outdir, "functions.tsv")
writeLines(render_markdown(functions), md_path, useBytes = TRUE)
write_tsv(tsv_path, functions)
message("wrote ", md_path)
message("wrote ", tsv_path)
